openfe/protocols/openmm_utils/multistate_analysis.py

```diff
@@ -0,0 +1,619 @@
+
+        Parameters
+        ----------
+        forward_reverse_samples : int
+          Number of samples to take in the foward and reverse analysis of
+          the free energies.
+        """
+        # Do things that get badly cached later
+        self._replica_states = self.analyzer.reporter.read_replica_thermodynamic_states()
+        self._equil_iters = self.analyzer.n_equilibration_iterations
+        self._prod_iters = self.analyzer._equilibration_data[2]
+
+        # Gather estimate of free energy
+        self._free_energy, self._free_energy_err = self.get_equil_free_energy()
+
+        # forward and reverse analysis
+        self._forward_reverse = self.get_forward_and_reverse_analysis(
+            forward_reverse_samples
+        )
+
+        # Gather overlap matrix
+        self._overlap_matrix = self.get_overlap_matrix()
+
+        # Gather exchange transition matrix
+        # Note we only generate these for replica exchange calculations
+        # TODO: consider if this would also work for SAMS
+        if self.sampling_method == 'repex':
+            self._exchange_matrix = self.get_exchanges()
+
+    @staticmethod
+    def _get_free_energy(
+        analyzer: multistate.MultiStateSamplerAnalyzer,
+        u_ln: npt.NDArray, N_l: npt.NDArray,
+        return_units: unit.Quantity,
+    ) -> tuple[unit.Quantity, unit.Quantity]:
+        """
+        Helper method to create an MBAR object and extract free energies
+        between end states.
+
+        Parameters
+        ----------
+        analyzer : multistate.MultiStateSamplerAnalyzer
+          MultiStateSamplerAnalyzer to extract free eneriges from.
+        u_ln : npt.NDArray
+          A n_states x (n_sampled_states * n_iterations)
+          array of energies (in kT).
+        N_l : npt.NDArray
+          An array containing the total number of samples drawn from each
+          state.
+        unit_type : unit.Quantity
+          What units to return the free energies in.
+
+        Returns
+        -------
+        DG : unit.Quantity
+          The free energy difference between the end states.
+        dDG : unit.Quantity
+          The MBAR error for the free energy difference estimate.
+
+        TODO
+        ----
+        * Allow folks to pass in extra options for bootstrapping etc..
+        * Add standard test against analyzer.get_free_energy()
+        """
+        mbar = analyzer._create_mbar(u_ln, N_l)
+
+        try:
+            # pymbar 3
+            DF_ij, dDF_ij = mbar.getFreeEnergyDifferences()
+        except AttributeError:
+            r = mbar.compute_free_energy_differences()
+            DF_ij = r['Delta_f']
+            dDF_ij = r['dDelta_f']
+
+        DG = DF_ij[0, -1] * analyzer.kT
+        dDG = dDF_ij[0, -1] * analyzer.kT
+
+        return (ensure_quantity(DG, 'openff').to(return_units),
+                ensure_quantity(dDG, 'openff').to(return_units))
+
+    def get_equil_free_energy(self) -> tuple[unit.Quantity, unit.Quantity]:
```

Comment:
how is this different from here: https://github.com/OpenFreeEnergy/openfe/blob/main/openfe/protocols/openmm_rfe/equil_rfe_methods.py#L646

Reply:
The way the method works in openmmtools currently is too restrictive, this is kind of a placeholder to eventually pass through bootstrapping etc.. as options.